### PR TITLE
refactor: migrate RSS3 API from v0.x to v1.x

### PR DESCRIPTION
**Pull request by UncleBill**
https://mask.atlassian.net/browse/MF-1837

### CODE REVIEW

File: packages/plugins/RSS3/src/constants.ts [hunk 1]
```diff
@@ -1,10 +1,9 @@
...
-import { resolveIPFSLink } from '@masknet/web3-shared-base'
 
 export const PLUGIN_ID = PluginId.RSS3
 export const PLUGIN_DESCRIPTION =
     'Derived from the best out of RSS, RSS3 is an open protocol designed for all our cyber existence in the era of Web 3.0.'
 export const PLUGIN_NAME = 'RSS3'
 
 /* cspell:disable-next-line */
-export const RSS3_DEFAULT_IMAGE = resolveIPFSLink('QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5')
+export const RSS3_DEFAULT_IMAGE = 'https://ipfs.io/ipfs/QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5'
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/constants.ts]: Why? It should return a link like this one.

File: packages/web3-providers/src/rss3/index.ts [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
-import { ChainId, SchemaType } from '@masknet/web3-shared-evm'
-import { NETWORK_PLUGIN, NEW_RSS3_ENDPOINT, RSS3_ENDPOINT, RSS3_FEED_ENDPOINT, TAG, TYPE } from './constants'
-import { NonFungibleTokenAPI, RSS3BaseAPI } from '../types'
-import { fetchJSON } from '../helpers'
 import {
     createIndicator,
     createPageable,
     HubOptions,
-    leftShift,
+    // leftShift,
```
Comment (guanbinrui) [inline packages/web3-providers/src/rss3/index.ts]:
```suggestion
```

File: packages/web3-shared/evm/src/types/index.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export type ChainIdRecord<T> = { [k in ChainId]: T }
 export enum ChainId {
     // Mainnet
     Mainnet = 1,
+    Classic = 61,
```
Comment (guanbinrui) [inline packages/web3-shared/evm/src/types/index.ts]: Adding new chains here will cause a chain reaction.

File: packages/web3-providers/src/types/RSS3.ts [hunk 4]
```diff
@@ -58,41 +68,240 @@ export namespace RSS3BaseAPI {
...
+        action: 'add' | 'remove'
+        tokens: [TransactionMetadata, TransactionMetadata]
+    }
+
+    interface CollectibleMetadata {
+        id: string
+        value: string
+        value_display: string
+        contract_address: string
+        standard: 'ERC-721' | string
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/RSS3.ts:118]:
```suggestion
        standard: 'ERC-721' | Omit<string, 'ERC-721'>
```

File: packages/web3-providers/src/rss3/index.ts [hunk 5]
```diff
@@ -46,25 +37,34 @@ export class RSS3API implements RSS3BaseAPI.Provider, NonFungibleTokenAPI.Provid
...
+        const { result: donations } = await fetchJSON<{ result: RSS3BaseAPI.Donation[] }>(collectionURL)
+        // A donation Feed contains multiple donation Actions. Let's flatten them.
+        const result = donations
+            .map((donation) => {
+                return donation.actions.map((action) => ({
+                    ...donation,
+                    actions: [action],
+                }))
+            })
+            .flat()
```
Comment (guanbinrui) [inline packages/web3-providers/src/rss3/index.ts]:
```suggestion
            .flatMap((donation) => {
                return donation.actions.map((action) => ({
                    ...donation,
                    actions: [action],
                }))
            })
```

File: packages/web3-providers/src/types/RSS3.ts [hunk 6]
```diff
@@ -140,7 +361,7 @@ export namespace RSS3BaseAPI {
         imageURL?: string
         title?: string
         description?: string
-        actions?: Action[]
+        // actions?: Action[]
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/RSS3.ts]:
```suggestion
```

File: packages/plugins/RSS3/src/SNSAdaptor/hooks/useDonations.ts [hunk 7]
```diff
@@ -1,10 +1,8 @@
...
 import { useAsync } from 'react-use'
-import type { AsyncState } from 'react-use/lib/useAsync'
-import { RSS3, RSS3BaseAPI } from '@masknet/web3-providers'
 
-export function useDonations(address: string): AsyncState<RSS3BaseAPI.Collection[] | undefined> {
+export function useDonations(address: string) {
     return useAsync(async () => {
-        const response = await RSS3.getDonations(address)
-        return response
+        return RSS3.getDonations(address)
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/SNSAdaptor/hooks/useDonations.ts]: Does it necessary to query with a `chainId`?

File: packages/plugins/RSS3/src/SNSAdaptor/hooks/useFootprints.ts [hunk 8]
```diff
@@ -1,8 +1,7 @@
-import { RSS3, RSS3BaseAPI } from '@masknet/web3-providers'
+import { RSS3 } from '@masknet/web3-providers'
 import { useAsync } from 'react-use'
-import type { AsyncState } from 'react-use/lib/useAsync'
 
-export function useFootprints(address: string): AsyncState<RSS3BaseAPI.Collection[] | undefined> {
+export function useFootprints(address: string) {
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/SNSAdaptor/hooks/useFootprints.ts:4]: Does it necessary to query with a `chainId`?

File: packages/plugins/RSS3/src/SNSAdaptor/components/FeedCard.tsx [hunk 9]
```diff
@@ -171,45 +161,78 @@ export const FeedCard = memo(({ feed, address, onSelect }: FeedCardProps) => {
...
-            )
+        if (feed.tag === Tag.Donation) {
+            const action = feed.actions[0] as RSS3BaseAPI.ActionGeneric<RSS3BaseAPI.Tag.Donation>
+            const logo = action.metadata?.logo
+            return <img className={classes.img} src={logo} />
         }
         return null
     }, [feed])
 
+    const normalizedFeed: RSS3Feed = useMemo(() => {
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/SNSAdaptor/components/FeedCard.tsx]: Better not to include a big chunk of data progressing in a UI component.

File: packages/plugins/RSS3/src/SNSAdaptor/components/DonationCard.tsx [hunk 10]
```diff
@@ -113,10 +114,10 @@ export const DonationCard = memo(({ donation, socialAddress, onSelect, className
                     <Typography className={classes.activity}>
                         <span className={classes.fontColor}>{reversedAddress}</span>{' '}
                         <span className={classes.fontColor}>{t.contributed()}</span>{' '}
-                        <span className={classes.tokenInfoColor}>{donation.tokenAmount?.toString()}</span>
-                        <span className={classes.tokenInfoColor}>{donation.tokenSymbol ?? 'ETH'}</span>{' '}
+                        <span className={classes.tokenInfoColor}>{action.metadata?.token.value_display}</span>
+                        <span className={classes.tokenInfoColor}>{action.metadata?.token.symbol ?? 'ETH'}</span>{' '}
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/SNSAdaptor/components/DonationCard.tsx]: If the `symbol` doesn't exist, then we can omit a related chunk of UI.

File: packages/plugins/RSS3/src/constants.ts [hunk 11]
```diff
@@ -7,4 +7,4 @@ export const PLUGIN_DESCRIPTION =
 export const PLUGIN_NAME = 'RSS3'
 
 /* cspell:disable-next-line */
-export const RSS3_DEFAULT_IMAGE = resolveIPFSLink('QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5')
+export const RSS3_DEFAULT_IMAGE = resolveIPFSLink('QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5') as string
```
Comment (guanbinrui) [inline packages/plugins/RSS3/src/constants.ts]:
```suggestion
export const RSS3_DEFAULT_IMAGE = resolveIPFSLink('QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5')!
```

File: packages/web3-providers/src/rss3/index.ts [hunk 12]
```diff
@@ -115,44 +115,19 @@ export class RSS3API implements RSS3BaseAPI.Provider, NonFungibleTokenAPI.Provid
...
             network: NETWORK_PLUGIN[networkPluginId],
             limit: 100,
-            exclude_tags: TAG.POAP,
-            latest: false,
+            include_poap: true,
         })
-        const res = fetchJSON<RSS3BaseAPI.Web3FeedResponse>(url + '&tags=Gitcoin&tags=POAP&tags=NFT&tags=Donation')
-        return res
+        const res = await fetchJSON<{ result: RSS3BaseAPI.Activity[] }>(url)
+        return res.result
```
Comment (guanbinrui) [inline packages/web3-providers/src/rss3/index.ts]: Does this API support return with a pagination?

File: packages/web3-providers/src/rss3/index.ts [hunk 13]
```diff
@@ -115,44 +115,19 @@ export class RSS3API implements RSS3BaseAPI.Provider, NonFungibleTokenAPI.Provid
...
+    /**
+     * Get feeds in tags of donation, collectible and transaction
+     */
+    async getWeb3Feeds(address: string, networkPluginId = NetworkPluginID.PLUGIN_EVM) {
+        if (!address) return []
+        const tags = [RSS3BaseAPI.Tag.Donation, RSS3BaseAPI.Tag.Collectible]
+        const url = urlcat(NEW_RSS3_ENDPOINT, `/:address?tag=${tags.join('&tag=')}`, {
             address,
             network: NETWORK_PLUGIN[networkPluginId],
             limit: 100,
```
Comment (guanbinrui) [inline packages/web3-providers/src/rss3/index.ts]: Should we promote it as a size property?

File: packages/web3-providers/src/types/RSS3.ts [hunk 14]
```diff
@@ -58,41 +68,240 @@ export namespace RSS3BaseAPI {
...
         platform?: PLATFORM
         related_urls?: string[]
     }
 
+    export interface AddressStatus {
+        update_at: RFC3339Datetime
+        address: string
+    }
+
     export type NETWORK =
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/RSS3.ts]: A wired name by comparing with other ones.

File: packages/web3-providers/src/types/RSS3.ts [hunk 15]
```diff
@@ -58,41 +68,240 @@ export namespace RSS3BaseAPI {
...
+        zksync: ChainId.ZKSync_Alpha_Testnet,
+        xdai: ChainId.xDai,
+        arbitrum: ChainId.Arbitrum,
+        optimism: ChainId.Optimism,
+        fantom: ChainId.Fantom,
+        avalanche: ChainId.Avalanche,
+        crossbell: ChainId.Crossbell,
+    }
+
     export type PLATFORM = 'mirror' | 'lens' | 'gitcoin' | 'snapshot' | 'uniswap' | 'binance' | 'crossbell'
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/RSS3.ts]:
```suggestion
    export type ActionPlatform = 'mirror' | 'lens' | 'gitcoin' | 'snapshot' | 'uniswap' | 'binance' | 'crossbell'
```

File: packages/web3-providers/src/types/RSS3.ts [hunk 16]
```diff
@@ -58,41 +68,240 @@ export namespace RSS3BaseAPI {
...
+    export enum Tag {
+        Collectible = 'collectible',
+        Donation = 'donation',
+        Exchange = 'exchange',
+        Governance = 'governance',
+        Social = 'social',
+        Transaction = 'transaction',
+    }
 
     export type TYPE =
```
Comment (guanbinrui) [inline packages/web3-providers/src/types/RSS3.ts]: The following one replaces it, isn't it?

File: packages/plugins/RSS3/src/SNSAdaptor/hooks/useDonations.ts (hunk 7 above)
Comment (UncleBill) [reply packages/plugins/RSS3/src/SNSAdaptor/hooks/useDonations.ts]: It's not according to the original code.

File: packages/plugins/RSS3/src/SNSAdaptor/hooks/useFootprints.ts (hunk 8 above)
Comment (UncleBill) [reply packages/plugins/RSS3/src/SNSAdaptor/hooks/useFootprints.ts:4]: ditto.

File: packages/web3-providers/src/types/RSS3.ts (hunk 15 above)
Comment (UncleBill) [reply packages/web3-providers/src/types/RSS3.ts]: feeds have `platform` as well.

File: packages/web3-providers/src/types/RSS3.ts (hunk 16 above)
Comment (UncleBill) [reply packages/web3-providers/src/types/RSS3.ts]: Indeed, it's unused.

File: packages/plugins/RSS3/src/SNSAdaptor/components/FeedCard.tsx (hunk 9 above)
Comment (UncleBill) [reply packages/plugins/RSS3/src/SNSAdaptor/components/FeedCard.tsx]: Abstract as a custom hook.

File: packages/web3-providers/src/rss3/index.ts [hunk 17]
```diff
@@ -115,44 +119,24 @@ export class RSS3API implements RSS3BaseAPI.Provider, NonFungibleTokenAPI.Provid
...
             network: NETWORK_PLUGIN[networkPluginId],
-            limit: 100,
-            exclude_tags: TAG.POAP,
-            latest: false,
+            include_poap: true,
         })
-        const res = fetchJSON<RSS3BaseAPI.Web3FeedResponse>(url + '&tags=Gitcoin&tags=POAP&tags=NFT&tags=Donation')
-        return res
+        const res = await fetchJSON<{ result: RSS3BaseAPI.Activity[]; cursor?: string }>(url)
+        return createPageable(res.result, createIndicator(indicator, res.cursor))
```
Comment (guanbinrui) [inline packages/web3-providers/src/rss3/index.ts]:
```suggestion
        return createPageable(res.result, createIndicator(indicator), createNextIndicator(indicator, res.cursor))
```

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**